pyartifactory/objects.py

```diff
@@ -1,684 +1,745 @@
+
+from pyartifactory.models import (
+    AuthModel,
+    ApiKeyModel,
+    PasswordModel,
+    AccessTokenModel,
+    Group,
+    LocalRepository,
+    VirtualRepository,
+    LocalRepositoryResponse,
+    VirtualRepositoryResponse,
+    RemoteRepository,
+    RemoteRepositoryResponse,
+    SimpleRepository,
+    UserResponse,
+    NewUser,
+    SimpleUser,
+    User,
+    Permission,
+    SimplePermission,
+    ArtifactPropertiesResponse,
+    ArtifactStatsResponse,
+)
+
+
+class Artifactory:
+    def __init__(
+        self,
+        url: str,
+        auth: Tuple[str, str] = None,
+        verify: bool = True,
+        cert: str = None,
+    ):
+        self.artifactory = AuthModel(url=url, auth=auth, verify=verify, cert=cert)
+        self.users = ArtifactoryUser(self.artifactory)
+        self.groups = ArtifactoryGroup(self.artifactory)
+        self.security = ArtifactorySecurity(self.artifactory)
+        self.repositories = ArtifactoryRepository(self.artifactory)
+        self.artifacts = ArtifactoryArtifact(self.artifactory)
+
+
+class ArtifactoryObject:
+    def __init__(self, artifactory: AuthModel) -> None:
+        self._artifactory = artifactory
+        self._auth = (
+            self._artifactory.auth[0],
+            self._artifactory.auth[1].get_secret_value(),
+        )
+        self._verify = self._artifactory.verify
+        self._cert = self._artifactory.cert
+        self.session = requests.Session()
+
+    def _get(self, route: str, **kwargs) -> Response:
+        """
+        :param route: API Route
+        :param kwargs: Additional parameters to add the request
+        :returns  An HTTP response
+        """
+        return self._generic_http_method_request("get", route, **kwargs)
+
+    def _post(self, route: str, **kwargs) -> Response:
+        """
+        :param route: API Route
+        :param kwargs: Additional parameters to add the request
+        :returns  An HTTP response
+        """
+        return self._generic_http_method_request("post", route, **kwargs)
+
+    def _put(self, route: str, **kwargs) -> Response:
+        """
+        :param route: API Route
+        :param kwargs: Additional parameters to add the request
+        :returns  An HTTP response
+        """
+        return self._generic_http_method_request("put", route, **kwargs)
+
+    def _delete(self, route: str, **kwargs) -> Response:
+        """
+        :param route: API Route
+        :param kwargs: Additional parameters to add the request
+        :returns  An HTTP response
+        """
+        return self._generic_http_method_request("delete", route, **kwargs)
+
+    def _generic_http_method_request(
+        self, method: str, route: str, **kwargs
+    ) -> Response:
+        """
+        :param method: HTTP method to use
+        :param route: API Route
+        :param kwargs: Additional parameters to add the request
+        :return: An HTTP response
+        """
+        raise_for_status = kwargs.pop("raise_for_status", True)
```

Comment:
I'm not sure about this feature, it would require rewriting all existing methods. I think that we should keep the current way of raising errors and if we want to modify it, another pull request should be made.

Reply:
actually I did it this way because it defaults to true, so it will not break your interface and will allow me to bypass raising for status when it is required. 

the reason I did this is because raising_for_status here supresses the real cause of the error and leaves you guessing. when a token can't be provisioned, If I don't pass on the response its ambiguous as to what is going on. 

example 

`art.security.create_access_token(user_name='bobby')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "/Users/ghawker/code/python/python-artifactory-fork/pyartifactory/objects.py", line 227, in create_access_token
    response.json().get("error_description", "Unknown error")
pyartifactory.exception.InvalidTokenDataException: Insufficient scope: 'api:*'`

here above we can see that a transient user gets created and you must pass groups or token generation doesn't work. with this call

`response = self._post(
            f"api/{self._uri}/token/revoke", data=payload, raise_for_status=False
        )`

I'm able to collect and handle the error correctly (it's useful for the user of the library). 

If I make the same call and pass no optional raise_for_status parameter like this 

`response = self._post(
            f"api/{self._uri}/token/revoke", data=payload
        )`
then you can see that the interface isn't broken (and no method rewriting needs to take place)

`raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: 'myurl.artifactory.com'`

of course this is your code and I can edit this out and submit as seperate PR if you like, but I do think this demonstrates both backwards compatability and a value addition for better/easier error handling.

Reply:
It's actually a great feature, thanks for your explanation :ok_hand: . As you said, it can be useful to the users of the library and even think that I could use that in other methods :+1:

Reply:
I read your explanation @dogfish182, and I find the idea and implementation quite nice, but it feels weird to take it as a `kwargs` parameter. Since we know how to handle it and are not passing it to the actual HTTP call, I think it could be changed to a default parameter of the method. That way it would be better documented and discoverable, while still being optional.

That would change it to
```diff
     def _generic_http_method_request(
-        self, method: str, route: str, **kwargs
+        self, method: str, route: str, raise_for_status: bool = True, **kwargs
     ) -> Response:
         """
         :param method: HTTP method to use
         :param route: API Route
         :param kwargs: Additional parameters to add the request
         :return: An HTTP response
         """
-        raise_for_status = kwargs.pop("raise_for_status", True)
         http_method = getattr(self.session, method)
```